### DISCUSSION TITLE
Почему нет поддоменов в списках для dnsmasq

### DISCUSSION THREAD

**Discussion by Danstiv** [itdoginfo/allow-domains#28] [Q&A] [answered]
Приветствую!
В ридми написано следующее:
>Dnmasq работает по wildcard. Это означает, что при добавлении домена `domain.com`, в списки IP-адресов будут добавляться также все поддомены `subdomain.domain.com`.

И далее:

> Поэтому Dnsmasq списки состоят только из доменов второго уровня.

Мне непонятно, как это связано, т.е. из первой цитаты не следует, что dnsmasq будет обрабатывать домены второго уровня, даже если ему явно задать домен 3-го уровня.
Но из второй цитаты, как будто, следует именно это.
Я решил покурить мануал, и нашёл следующее.
Из описания nftset:
> Similar to the --ipset option, but accepts one or more nftables sets to add IP addresses into.

Окей, смотрим на ipset.

> Domains and subdomains are matched in the same way as --address

Читаем про address

> The domain specification works in the same way as for --server

Читаем про --server, и ура! Здесь это наконец-то документировано 🙂

> More specific domains take precedence over less specific domains, so: --server=/google.com/1.2.3.4 --server=/www.google.com/2.3.4.5 will send queries for google.com and gmail.google.com to 1.2.3.4, but www.google.com will go to 2.3.4.5
> Matching of domains is normally done on complete labels, so /google.com/ matches google.com and www.google.com but NOT supergoogle.com. This can be overridden with a * at the start of a pattern only: /*google.com/ will match google.com and www.google.com AND supergoogle.com. The non-wildcard form has priority, so if /google.com/ and /*google.com/ are both specified then google.com and www.google.com will match /google.com/ and /*google.com/ will only match supergoogle.com.

Так почему, например, нельзя вписать в dnsmasq-овский конфиг `nftset=/bard.google.com/4#inet#fw4#vpn_domains`?

Я попробовал, и ничего не ломается, bard.google.com маршрути... кхм. блокируется, а google.com и play.google.com пропускаются.

Так почему же нельзя субдомены использовать в dnsmasq, что я упускаю?

Заранее спасибо за ответ.

**Comment by Danstiv** [comment]
@itdoginfo
Упоминаю на всякий случай, будет круто, если сможете ответить.

**Comment by itdoginfo** [comment, marked as answer]
Вы правы, сейчас всё работает именно так, как вы описали. Dnsmasq обрабатывает субдомены, не обрабатывая основной домен.

Я всё это время ориентировался на старую логику, когда он обрабатывал все домены и субдомены, даже если был указан только один субдомен. Когда произошло это изменение не совсем понятно из их changelog.

Я изменил обработку скрипта b70825099b053e3b45b26429620df928a3ebba48 теперь в списки для dnsmasq также залетают субдомены отсюда https://github.com/itdoginfo/allow-domains/blob/main/src/Russia-domains-inside-single.lst

Спасибо за вашу проверку и что написали об этом.
